ns1_zone: add a new 'autogenerate_ns_record' parameter defaulting to 'true'

XaF opened the pull request:
This new parameters is only used on creating a zone and allows to
prevent for NS records to be created automatically, which would
prevent users to create those records by themselves. This does it
by actually removing the auto-created NS record when the option
is set to 'false'. By default, the option is set to 'true'. We
also ignore any change to that value after creation, as it will
not imply any change to the deployed resource.

Following the discussion we've had on https://github.com/terraform-providers/terraform-provider-ns1/pull/77

rupa commented on ns1/resource_zone.go:
```diff
@@ -279,11 +288,18 @@ func resourceDataToZone(z *dns.Zone, d *schema.ResourceData) {
 	client := meta.(*ns1.Client)
-	z := dns.NewZone(d.Get("zone").(string))
+	zone := d.Get("zone").(string)
+	z := dns.NewZone(zone)
```
this seems like the right approach, but my delicate sensibilities prefer leaving this line alone, and pulling the zoneName back out of `z` in the `if` clause.
not a very strong opinion and not based on anything i've measured, so if consensus is to keep it, i'd only ask we call it `zoneName` instead of `zone`

rupa commented on ns1/resource_zone.go:
```diff
@@ -120,6 +121,14 @@ func resourceZone() *schema.Resource {
 				Optional: true,
 				Computed: true,
 			},
+			"create_ns_record": {
```
we were talking about what to call this key, and were interested to see what you came up with :)

best stab i had was `autogenerate_ns_record`, I think it's clearer

XaF commented:
Addressed your comments @rupa - Had to force-push to replace the commit message to include the new option name!

I think the AccTest would be doable by creating a zone with that option, and then trying to grab the NS records, which should result in no existing record (as we just deleted it). It could be compared with the creation of another zone not using the option that would have the NS record in this case. This imply checking another resource in the AccTest however, not sure if this is very nice.

I'll continue my thought process about that. If you get better ideas I'm all for it :)

XaF commented:
Thoughts on the added test ?

mburtless commented on ns1/resource_zone.go:
```diff
@@ -284,6 +293,12 @@ func zoneCreate(d *schema.ResourceData, meta interface{}) error {
 	if _, err := client.Zones.Create(z); err != nil {
 		return err
 	}
+	if !d.Get("autogenerate_ns_record").(bool) {
```
This always evaluates `true` as `autogenerate_ns_record` does not appear to be saved in state.  NS records will always be deleted.

XaF replied:
Should not be an issue anymore, I removed the `DiffSuppressFunc`

XaF commented on ns1/resource_zone_test.go:
```diff
@@ -367,6 +391,43 @@ func testAccCheckZoneExists(n string, zone *dns.Zone) resource.TestCheckFunc {
+
+		p := rs.Primary
+
+		if expected != p.Attributes["autogenerate_ns_record"] {
```
Right! Will fix that

XaF commented on ns1/resource_zone_test.go:
```diff
@@ -367,6 +391,43 @@ func testAccCheckZoneExists(n string, zone *dns.Zone) resource.TestCheckFunc {
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("NoID is set")
```
Did not realize, but it's coming from a copy paste from https://github.com/terraform-providers/terraform-provider-ns1/blob/master/ns1/resource_record_test.go#L242 - Will fix it here but might need a typo fix in the other file!

mburtless replied:
Thanks for the heads up!  I'll get this corrected in the other spots.

XaF replied:
I fixed the other mention in the same file though, as it was just a few lines above

XaF replied:
Also, might be me but `No ID is set` seems weird to read, while `ID is not set` would feel more natural ?

mburtless replied:
I'm with you.  It is not a super descriptive error either way when you see it thrown in tests, in my opinion.  Probably something to revisit later.

XaF commented:
@mburtless I think I fixed the issues. Was able to run the AccTest locally and it works (tested the new one and the basic one that includes the 'NS is kept around' test)

XaF commented:
> `DiffSuppressFunc` was nice after initial create, but we can always add one back in later.

Forgot to mention it, but seems that `DiffSuppressFunc` has an issue with `Default`, and the later is ignored when the former is present, which is why I removed it. There is a issue talking about that: https://github.com/hashicorp/terraform-plugin-sdk/issues/70
